lib/internal/modules/esm/resolve.js

```diff
@@ -1,26 +1,612 @@
+  if (getOptionValue('--experimental-specifier-resolution') === 'node') {
+    let file = resolveExtensionsWithTryExactName(resolved);
+    if (file !== undefined) return file;
+    if (!StringPrototypeEndsWith(resolved.pathname, '/')) {
+      file = resolveIndex(new URL(`${resolved.pathname}/`, base));
+    } else {
+      file = resolveIndex(resolved);
+    }
+    if (file !== undefined) return file;
+    throw new ERR_MODULE_NOT_FOUND(
+      resolved.pathname, fileURLToPath(base), 'module');
+  }
+
+  if (StringPrototypeEndsWith(resolved.pathname, '/')) return resolved;
+  const path = fileURLToPath(resolved);
+
+  if (!tryStatSync(path).isFile()) {
+    throw new ERR_MODULE_NOT_FOUND(
+      path || resolved.pathname, fileURLToPath(base), 'module');
+  }
+
+  return resolved;
+}
+
+function throwExportsNotFound(subpath, packageJSONUrl, base) {
+  throw new ERR_PACKAGE_PATH_NOT_EXPORTED(
+    fileURLToPath(packageJSONUrl), subpath, fileURLToPath(base));
+}
+
+function throwSubpathInvalid(subpath, packageJSONUrl, base) {
+  throw new ERR_INVALID_MODULE_SPECIFIER(
+    fileURLToPath(packageJSONUrl), subpath, fileURLToPath(base));
+}
+
+function throwExportsInvalid(
+  subpath, target, packageJSONUrl, base) {
+  if (typeof target === 'object' && target !== null) {
+    target = JSONStringify(target, null, '');
+  } else if (ArrayIsArray(target)) {
+    target = `[${target}]`;
+  } else {
+    target = `${target}`;
+  }
+  throw new ERR_INVALID_PACKAGE_TARGET(
+    fileURLToPath(packageJSONUrl), null, subpath, target, fileURLToPath(base));
+}
+
+function resolveExportsTargetString(
+  target, subpath, match, packageJSONUrl, base) {
+  if (target[0] !== '.' || target[1] !== '/' ||
+      (subpath !== '' && target[target.length - 1] !== '/')) {
+    throwExportsInvalid(match, target, packageJSONUrl, base);
+  }
+
+  const resolved = new URL(target, packageJSONUrl);
+  const resolvedPath = resolved.pathname;
+  const packagePath = new URL('.', packageJSONUrl).pathname;
+
+  if (!StringPrototypeStartsWith(resolvedPath, packagePath) ||
+      StringPrototypeIncludes(
+        resolvedPath, '/node_modules/', packagePath.length - 1)) {
+    throwExportsInvalid(match, target, packageJSONUrl, base);
+  }
 
-function defaultResolve(specifier, { parentURL } = {}, defaultResolve) {
+  if (subpath === '') return resolved;
+  const subpathResolved = new URL(subpath, resolved);
+  const subpathResolvedPath = subpathResolved.pathname;
+  if (!StringPrototypeStartsWith(subpathResolvedPath, resolvedPath) ||
+      StringPrototypeIncludes(subpathResolvedPath,
+                              '/node_modules/', packagePath.length - 1)) {
+    throwSubpathInvalid(match + subpath, packageJSONUrl, base);
+  }
+  return subpathResolved;
+}
+
+function isArrayIndex(key) {
+  const keyStr = `${key}`;
+  if (`${+keyStr}` !== keyStr) return false;
+  key >>>= 0;
+  if (ObjectIs(key, 0)) return true;
+  return key > 0 && key < 0xFFFF_FFFF;
+}
```

Comment:
`key` is always a string coming in. @ljharb / @jkrems this function treats `"-0"` as "0" currently and an ArrayIndex, is that correct or is it supposed to complain about `"-0"`. We can simplify some of the coercions but if we only check for `0` not `-0` we should use `ObjectIs` instead of `===` below:

```suggestion
function isArrayIndex(key /* string */) { /* -> boolean */
  const keyNum = +key;
  if (`${keyNum}` !== key) return false;
  return keyNum === 0 || (keyNum > 0 && keyNum < 0xFFFF_FFFF);
}
```

Reply:
It seems reasonable to me to disallow both + and - zero

Reply:
+0 is a valid array index, though?

Reply:
So now that this is in JS, we can use a JS implementation reference.

I would suggest copying the spec correct implementation from https://github.com/engine262/engine262/blob/b8af8fec080b1f61c3e3ec6f10295f3187430b9c/src/abstract-ops/data-types-and-values.mjs#L34 verbatim here.

Reply:
Maybe I'm misunderstanding the purpose of this code; if it's trying to only ban strings that are valid array indices, then `-0` shouldn't be banned, but I don't see a use case for a `"-0"` entry point either.

Reply:
This is to handlehttps://tc39.es/ecma262/#sec-ordinaryownpropertykeys 's usage of https://tc39.es/ecma262/#array-index to confirm that numeric key ordering isn't surprising. Reading, it looks like we should only accept `+0`.

Reply:
FWIW, at least for V8's `Object.keys` neither of the keys `-0` and `+0` trigger a surprising order (and that's also how I'm reading the spec):

```console
> Object.keys({ x: 1, y: 2, 0: 3 })
[ '0', 'x', 'y' ]
> Object.keys({ x: 1, y: 2, '0': 3 })
[ '0', 'x', 'y' ]
> Object.keys({ x: 1, y: 2, '-0': 3 })
[ 'x', 'y', '-0' ]
> Object.keys({ x: 1, y: 2, '+0': 3 })
[ 'x', 'y', '+0' ]
```

The string `+0` itself isn't the canonical representation of its numeric value `0` so we don't have to worry about it I think.

Reply:
'+0' and '-0' the strings are false via `(`${keyNum}` !== key) return false;`

Reply:
i've updated the suggestion and it should be matching spec concern now

Reply:
> +0 is a valid array index, though?

I don't think so once if `+0` is the actual property key. I only works if `+0` gets interpreted as `0` and stringified to that value.

```
> [0]['-0']
undefined
> [0]['+0']
undefined
> [0]['0']
0
```

Reply:
Updated the PR with the updated suggestion :)

Reply:
LGTM